Readme.pod
=pod

=head1 NAME

Mojolicious::Plugin::PubSubHubbub - Publish and Subscribe with PubSubHubbub


=head1 SYNOPSIS

  # Mojolicious
  $app->plugin(PubSubHubbub => {
    hub => 'https://hub.example.org/',
    pubsub_verify => sub {
      return 1;
    }
  });

  my $r = $app->routes;
  $r->route('/:user/callback_url')->pubsub;

  # Mojolicious::Lite
  plugin 'PubSubHubbub' => {
    hub => 'https://hub.example.org'
  };

  any('/:user/callback_url')->pubsub;

  # In Controllers:
  # Publish feeds to subscribers
  $c->pubsub->publish(
    'https://sojolicio.us/blog.atom',
    'https://sojolicio.us/activity.atom'
  );

  # Subscribe to a feed
  $c->pubsub->subscribe(
    topic => 'https://sojolicio.us/feed.atom',
    hub   => 'https://hub.sojolicio.us'
  );

  # Discover a resource
  my ($topic, $hub) = $c->pubsub->discover('http://sojolicio.us/');
  if ($topic && $hub) {
    $c->pubsub->subscribe( topic => $topic, hub   => $hub );
  };

  # Unsubscribe from a feed
  $c->pubsub->unsubscribe(
    topic => 'https://sojolicio.us/feed.atom',
    hub   => 'https://hub.sojolicio.us'
  );


=head1 DESCRIPTION

L<Mojolicious::Plugin::PubSubHubbub> is a plugin to publish and subscribe to
L<PubSubHubbub 0.3|http://pubsubhubbub.googlecode.com/svn/trunk/pubsubhubbub-core-0.3.html>
Webhooks.

The plugin currently supports the publisher and subscriber part of the protocol,
I<not> the hub part.

This plugin is data store agnostic.
Please use this plugin by applying hooks and callbacks.

B<This module is an early release! There may be significant changes in the future.>


=head1 ATTRIBUTES

=head2 hub

  $ps->hub('http://pubsubhubbub.appspot.com/');
  my $hub = $ps->hub;

The preferred hub. Currently local hubs are not supported.
Establishes an L<endpoint|Mojolicious::Plugin::Util::Endpoint> called C<pubsub-hub>.

Defaults to L<pubsubhubbub.appspot.com|http://pubsubhubbub.appspot.com/>,
but this may change without notification.


=head2 lease_seconds

  my $seconds = $ps->lease_seconds;
  $ps->lease_seconds(100 * 24 * 60 * 60);

Seconds a subscription is valid by default before auto refresh
is enabled. You can not rely on auto refresh by the hub -
your subscriber should resubscribe automatically, if the hub
does not ask for renewal.

Defaults to 9 days.


=head1 METHODS

=head2 register

  # Mojolicious
  $app->plugin(PubSubHubbub => {
    hub => 'https://hub.example.org/',
    lease_seconds => 100 * 24 * 60 * 60
  });

  # Mojolicious::Lite
  plugin 'PubSubHubbub' => {
    hub => 'https://hub.example.org/',
    lease_seconds => 100 * 24 * 60 * 60
  };

  # Or in your config file
  {
    PubSubHubbub => {
      hub => 'https://hub.example.org/',
      lease_seconds => 100 * 24 * 60 * 60
    }
  }


Called when registering the plugin.
Accepts the attributes mentioned as parameters.

All parameters can be set either as part of the configuration
file with the key C<PubSubHubbub> or on registration
(that can be overwritten by configuration).


=head1 SHORTCUTS

=head2 pubsub

  # Mojolicious
  my $r = $app->routes;
  $r->route('/callback_url')->pubsub;

  # Mojolicious::Lite
  any('/callback_url')->pubsub;

Define the callback endpoint for your subscriptions.
Establishes an L<endpoint|Mojolicious::Plugin::Util::Endpoint>
called C<pubsub-callback>.


=head1 HELPERS

=head2 pubsub->discover

  # In Controllers
  my ($topic, $hub) = $c->pubsub->discover('http://sojolicio.us/');

Discover a topic feed and a hub based on a URI.
The discovery heuristics may change without notification.


=head2 pubsub->publish

  # In Controllers
  my $success = $c->pubsub->publish(
    'my_feed',                       # named route
    '/feed.atom',                    # relative paths
    'https://sojolicio.us/feed.atom' # absolute URIs
  );

Publish a list of feeds in terms of a notification to the hub.
Supports endpoints, named routes, relative paths and absolute URIs.
Returns a true value on success.


=head2 pubsub->subscribe

  # In Controllers
  if ($c->pubsub->subscribe(
    topic => 'https://sojolicio.us/feed.atom',
    hub   => 'https://hub.sojolicio.us',
    lease_seconds => 123456
  )) {
    print 'You successfully subscribed!';
  };

Subscribe to a topic.

Relevant parameters are C<hub>,
C<lease_seconds>, C<secret>, C<verify_token>, and C<callback>.
Additional parameters are ignored but can be accessed in the hooks.
If no C<verify_token> is given, it is automatically generated.
If no C<callback> is given, the route callback is used.
If no C<lease_seconds> is given, the subscription won't
automatically terminate.
If a C<secret> is given, it must be unique for every C<callback>
and C<hub> combination to allow for bulk distribution.

The method returns a C<true> value on success and a C<false> value
if an error occured. If called in an array context, the
hub's response message body is returned additionally.


=head2 pubsub->unsubscribe

  # In Controllers
  if ($c->pubsub->unsubscribe(
    topic => 'https://sojolicio.us/feed.atom',
    hub   => 'https://hub.sojolicio.us'
  )) {
    print 'You successfully unsubscribed!';
  };

Unsubscribe from a topic.

Relevant parameters are C<hub>, C<secret>, C<verify_token>, and C<callback>.
Additional parameters are ignored but can be accessed in the hooks.
If no C<verify_token> is given, it is automatically generated.
If no C<callback> is given, the route callback is used.

The method returns a C<true> value on success and a C<false> value
if an error occured. If called in an array context, the
hub's response message body is returned additionally.


=head1 CALLBACKS

=head2 pubsub_accept

  # Establish callback
  $app->callback(
    pubsub_accept => sub {
      my ($c, $type, $topics) = @_;

      # Filter topics
      my @new_topics = grep($_ !~ /catz/, @$topics);

      # Set secret
      my $secret     = 'z0idberg';

      # Set X-Hub-On-Behalf-Of value
      my $on_behalf  = 3;
      return (\@new_topics, $secret, $on_behalf);
    });

This callback is released, when content arrives at the
pubsub endpoint. The parameters passed to the callback
include the current controller object, the content type,
and an array reference of topics.

Expects an array reference of maybe filtered topics,
a secret if necessary, and the value of C<X-Hub-On-Behalf-Of>.
If the returned topic list is empty, the processing will stop.
If the callback is not established, the complete content will be
processed.

The callback can be established with the
L<callback|Mojolicious::Plugin::Util::Callback/callback>
helper or on registration.


=head2 pubsub_verify

  # Establish callback
  $app->callback(
    pubsub_verify => sub {
      my ($c, $param) = @_;

      # Topic is valid
      if ($param->{topic} =~ /catz/ &&
          $param->{verify_token} eq 'zoidberg') {
        return 1;
      };

      # Not verified
      return;
    });

This callback is released, when a verification is requested.
The parameters include the current controller object and the parameters
of the verification request as a hash reference (without C<hub.>-prefix).
If verification is granted, this callback must return a true value.

The callback can be established with the
L<callback|Mojolicious::Plugin::Util::Callback/callback>
helper or on registration.


=head1 HOOKS

=head2 on_pubsub_content

  $app->hook(
    on_pubsub_content => sub {
      my ($c, $type, $dom) = @_;

      if ($type eq 'atom') {
        $dom->find('entry')->each(
          print $_->at('title')->text, "\n";
        );
      };

      return;
    });

This hook is released, when desired (i.e., verified and optionally
filtered) content arrives.
The parameters include the current
controller object, the content type (either C<atom> or C<rss>),
and the - maybe topic filtered - content as a L<Mojo::DOM> object.

The L<Mojo::DOM> object is canonicalized in a way that each
entry in the feed (either RSS or Atom) includes its topic in the C<href>
of C<source E<gt> link[rel="self"]>.


=head2 before_pubsub_subscribe

  $app->hook(
    before_pubsub_subscribe => sub {
      my ($c, $params, $post) = @_;

      my $topic = $params->{topic};
      print "Start following $topic\n";

      return;
    });

This hook is released, before a subscription request is sent to a hub.
The parameters include the current controller object,
the parameters prepared for subscription as a hash reference and the C<POST>
string as a string reference.
This hook can be used to store subscription information and establish
a secret.


=head2 after_pubsub_subscribe

  $app->hook(
    after_pubsub_subscribe => sub {
      my ($c, $hub, $params, $status, $body) = @_;
      if ($status !~ /^2/) {
        warn 'Error: ', $body;
      };

      return;
    });

This hook is released, after a subscription request is sent to a hub
and the response is processed.
The parameters include the current controller object,
the hub location,
the parameters sent for subscription as a hash reference, the response status,
and the response body.
This hook can be used to deal with errors.


=head2 before_pubsub_unsubscribe

  $app->hook(
    before_pubsub_unsubscribe => sub {
      my ($c, $params, $post) = @_;

      my $topic = $params->{topic};
      print "Stop following $topic\n";

      return;
    });

This hook is released, before an unsubscription request is sent
to a hub.
The parameters include the current controller object,
the parameters prepared for unsubscription as a hash reference and the C<POST>
string as a string reference.
This hook can be used to store unsubscription information.


=head2 after_pubsub_unsubscribe

  $app->hook(
    after_pubsub_unsubscribe => sub {
      my ($c, $hub, $params, $status, $body) = @_;
      if ($status !~ /^2/) {
        warn 'Error: ', $body;
      };

      return;
    });

This hook is released, after an unsubscription request is sent to a hub
and the response is processed.
The parameters include the current controller object,
the hub location,
the parameters sent for unsubscription as a hash reference, the response status,
and the response body.
This hook can be used to deal with errors.


=head1 EXAMPLE

The C<examples/> folder contains a full working example application with publishing,
subscription and discovery logic.
The example has additional dependencies of L<DBI>, L<DBD::SQLite> and
L<XML::Loy> (at least v0.13).

It can be started using the daemon, morbo or hypnotoad,
and needs to be accessible from the web.

  $ perl examples/pubsubapp daemon

=for HTML <br /><div style="text-align: center;"><img src="http://sojolicio.us/images/pubsubhubbub-screenshot.png" alt="PubSubHubbub Example Application" /></div>

This example may be a good starting point for your own implementation, especially,
if you deal with the subscriber part.

=head1 TODO

Currently all methods are blocking. In an upcoming release all blocking
methods will allow for non-blocking as well.


=head1 DEPENDENCIES

L<Mojolicious> (best with SSL support),
L<Mojolicious::Plugin::Util::Endpoint>,
L<Mojolicious::Plugin::Util::Callback>,
L<Mojolicious::Plugin::Util::RandomString>.


=head1 AVAILABILITY

  https://github.com/Akron/Mojolicious-Plugin-PubSubHubbub

This plugin is part of the L<Sojolicious|http://sojolicio.us> project.


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011-2018, L<Nils Diewald|http://nils-diewald.de/>.

This program is free software, you can redistribute it
and/or modify it under the terms of the Artistic License version 2.0.

=cut


Changes
0.19 2018-11-01
        - Change permissions for files (manwar++).

0.18 2017-12-27
        - Improved test coverage.
        - Fixed tests for recent Mojolicious versions.

0.17 2017-01-10
        - Updated to Mojolicious 7.13.
        - Removed deprecated methods pubsub_publish, pubsub_subscribe,
          pubsub_unsubscribe, pubsub_discover.

0.16 2016-07-26
        - Updated to Mojolicious 7.
        - Added pubsub->publish helper.
        - Added pubsub->subscribe helper.
        - Added pubsub->unsubscribe helper.
        - Added pubsub->discover helper.
        - Deprecated pubsub_publish in favor of pubsub->publish.
        - Deprecated pubsub_subscribe in favor of pubsub->subscribe.
        - Deprecated pubsub_unsubscribe in favor of pubsub->unsubscribe.
        - Deprecated pubsub_discover in favor of pubsub->discover.

0.15 2016-02-29
        - Updated to Mojolicious 6.48.

0.14 2015-04-24
        - Updated to Mojolicious 6.00.

0.13 2014-10-17
        - Updated to Mojolicious 5.49.
        - Fixed test suite for new every_param() method.

0.12 2014-05-31
        - Compatibility fix for link headers in Mojo versions > 4.99.

0.11 2014-05-29
        - Cleanup.
        - Fixed License issue (now coherent Artistic License 2.0).

0.10 2013-08-03
        - Update to Mojolicious 4.23.

0.09 2013-05-20
        - Fixed for Mojolicious 4.0 compatibility.

0.08 2013-04-19
        - Improved documentation.
        - Updated dependencies.

0.07 2013-03-31
        - Bug fixes from development versions.

0.06 2013-03-29
        - Improved documentation.
        - Improved content handling in example app.
        - Fixed bug in using "values" with a dereferenced array.
        - Fixed prototype bug.
        - 0.06_1: Improved test suite.
        - 0.06_2: Deleted all ~~ from test suite.
        - 0.06_3: Sort topics.

0.05 2013-03-23
        - Fixed a bug in the subscription process
          of the example app.

0.04 2013-03-22
        - Fixed base bug in discovery.
        - Fixed bug in encoding.
        - Made example app working in real world.
        - Don't use constants.
        - Improve documentation.
        - Change lease seconds default to 9 days.
        - Add screenshot of example app.
        - First CPAN release.

0.03 2013-03-14
        - Added discovery.
        - Improved example script.
        - Changed default lease seconds to 7 days,
          as appspot has a maximum of 10 days.

0.02 2013-03-12
        - Changed 'on_pubsub_acceptance' hook
          to 'pubsub_accept' callback.
        - Changed 'on_pubsub_verification' hook
          to 'pubsub_verify' callback.
        - Added example script.

0.01 2013-02-05
        - Small improvements to documentation.
        - Separation from Sojolicious package.
        - Initial submission to GitHub.


Makefile.PL
#!/usr/bin/env perl
use 5.010001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME         => 'Mojolicious::Plugin::PubSubHubbub',
  AUTHOR       => 'Nils Diewald',
  ABSTRACT     => 'Publish and Subscribe with PubSubHubbub',
  VERSION_FROM => 'lib/Mojolicious/Plugin/PubSubHubbub.pm',
  BUILD_REQUIRES => {
    'Test::More' => 0
  },
  PREREQ_PM => {
    'Mojolicious' => 6.00,
    'Mojolicious::Plugin::Util::Endpoint' => 0.19,
    'Mojolicious::Plugin::Util::Callback' => 0.02,
    'Mojolicious::Plugin::Util::RandomString' => 0.05
  },
  LICENSE      => 'artistic_2',
  META_MERGE   => {
    requires  => {perl => '5.010001'},
    resources => {
      license     => 'http://www.opensource.org/licenses/artistic-license-2.0',
      repository  => 'https://github.com/Akron/Mojolicious-Plugin-PubSubHubbub'
    }
  },
  MIN_PERL_VERSION => '5.010001',
  test => {TESTS => 't/*.t t/examples/*.t'}
);


sandboxes/PubSubHubbub/Hub.pm
# Do not accept negative values for X-Hub-On-Behalf-Of!


sub hub {
    my $self = shift;
    my $c = shift;
    my $param = $c->param;
    my %hub_param;

    foreach my $key (map(s/^hub\.//, grep('hub.', keys %$param))) {
	$hub_param{ $key } = $param->{ 'hub.' . $key };
    };

    my $mode = $hub_param{mode};

    my $ok = 0;

    # Subscribe to feed
    if ($mode eq 'subscribe') {
	if (exists $hub_param{ callback } &&
	    exists $hub_param{ topic } &&
	    ) {

	    # todo: topic: delete fragments.
            # from Google's hub (see subfeedr)
	    # my %valid_ports = map { $_ => 1 } (80, 443, 4443, 8080 .. 8089, 8188, 8440, 8990);	    

	    # Delete Whitespace from secret!
	    if (exists $hub_param{secret}) {
		trim($hub_param{secret});
	    };

	    $self->app->run_hook( 'on_pubsub_hub_subscription' => 
				  $self,
				  $c,
				  \%hub_param,
				  \$ok );
	};

	if ($$ok) {
	    my $verify_token = $param_hub{verify_token};


	    # VERIFY REQUEST
	    my $verified = 0;
	    my $challenge = _generate_challenge(4);

	    # Verification mode -- only 'sync' supported
#	    if ($hub_param{verify} eq 'sync' ||
#		$hub_param{verify} eq 'async') {

	    # Create verification request
	    my $verify_url = Mojo::URL->new($param_hub{callback});
	    my $params = $verify_url->params;
	    $params->append(
		'hub.mode'      => $param_hub{mode},
		'hub.topic'     => $param_hub{topic},
		'hub.challenge' => $challenge
		);

	    # Append 'verify_token' if existing
	    for ('verify_token') {
		$param->append('hub.'.$_ => $param_hub{$_})
		if exists $param_hub{$_};
	    };

	    # Append 'lease_seconds', which maybe is unset
	    for ('lease_seconds') {
		$param->append('hub.'.$_ =>
			       $param_hub{$_} || $self->lease_seconds );
	    };
	    
	    # Start request
	    my $ua = Mojo::UserAgent->new(max_redirects => 3)
	    my $get = $ua->get($verify_url->abs);
	    
	    if ($get->res->is_status_class(200)) {
		if ($get->res->body eq $challenge) {
		    # Verified!
		    $verified = 1;
		};
	    } elsif ($get->res->status == 404) {
		# definite fail
	    }

# Todo - wenn async kann ein fail nur temporär sein und sollte retried werden


#	    }

	    if ($hub_param{secret}) {
		# todo
	    };

	    # Request verified
	    if ($verified) {
		return $c->render(status => 204);
	    }

	    # Request not yet verified
	    elsif ($hub_param{verify} eq 'async') {
		return $c->render(status => 202)
	    };

	}

	# Not okay
	else {
	    $c->render(status => 404,
		       data   => 'This feed is not published '.
		                 'by this hub.' );
	};


    } elsif ($mode eq 'unsubscribe') {
    } elsif ($mode eq 'publish') {
    } else {
	# unknown mode
    };
};

# Verify a changed subscription or automatically refresh
sub hub_verify {
};

sub hub_publish {
    my $self = shift;
    my $c = shift;
    my $feed_url = shift;
    my $param = ref($_[0]) ? shift(@_) : {};
    my $payload = shift;

    # Subscriber structure:
    #   [{secret => 'ggfhgfhg'?, callback => 'http...'}*]

    my $subscribers = [];
    $self->app->run_hook('on_pubsub_hub_publishing' =>
			 $self,
			 $c,
			 $feed_url,
			 $subscribers);
    
    $param->{'Content-Type'} ||= 'application/atom+xml';

    my $ua = Mojo::UserAgent->new(max_redirects => 3)

    foreach my $subscriber (@$subscribers) {
	delete $param->{'X-Hub-Signature'};

	# Signature needed
	if (exists $subscriber->{secret}) {
	    # Todo: Check if this is 40char hexadecimal!
	    $param->{'X-Hub-Signature'} =
		'sha1=' . b($payload)->hmac_sha1_sum($subscriber->{secret});
	}

	my $response = $ua->post($subscriber->{callback} =>
				 $param =>
				 $payload);

	if ($response->res->is_status_class(200)) {
	    # ready
	} elsif ($response->res->code == 404) {
	    # fail
	} else {
	    # retry
	};
	
    };
};


sub _generate_challenge {
    my $length = shift;

    my $challenge = '';

    my $c_l = @challenge_chars - 1;
    foreach (0 .. $length) {
	$challenge .= $challenge_chars[int(rand($c_l))];
    };
    return $challenge;
};

1;

=pod

=head1 NAME

Mojolicious::Plugin::PubSubHubbub

=cut


my $ua = Mojo::UserAgent->new(max_redirects => 3);

foreach my $subscriber (@$subscribers) {
  delete $param->{'X-Hub-Signature'};

  # Signature needed
  if (exists $subscriber->{secret}) {
    # Todo: Check if this is 40char hexadecimal!
    $param->{'X-Hub-Signature'} =
      'sha1=' . b($payload)->hmac_sha1_sum($subscriber->{secret});
  }

  $ua->post(
    $subscriber->{callback} =>
      $param =>
	$payload => sub {
	  my ($ua, $tx) = @_;

	  my $host = $tx->req->host;

	  if ($tx->res->is_status_class(200)) {
	    # ready
	  } elsif ($tx->res->code == 404) {
	    # fail
	  } else {
	    # retry
	  };
	});

	
    };


examples/PubSubApp.pl
#!/usr/bin/env perl

# Todo: Add simple Blogging-Tool!
# Use XML::Loy!

use File::Basename 'dirname';
use File::Spec;
BEGIN {
  my @libdir = File::Spec->splitdir(dirname(__FILE__));
  use lib join '/', @libdir, 'lib';
  use lib join '/', @libdir, '..', 'lib';
};
use Mojolicious::Lite;
use Mojo::ByteStream 'b';
use Mojo::Collection 'c';
use Mojo::Date;
use DBI;
use DBD::SQLite;
use XML::Loy::Date::RFC3339;


# Maybe initialize the database
my $file = app->home . '/pubsub.sqlite';
sub _init_file;
unless (-e $file) {
  die 'Unable to init database' unless _init_db($file);
};


# Create database handler helper
helper dbh => sub {
  state $dbi = DBI->connect("dbi:SQLite:dbname=" . $file, "", "");
};


# Register pubSubHubbub plugin
plugin 'PubSubHubbub';


# Add 'acceptance' callback
app->callback(pubsub_accept => sub {
  my ($c, $type, $topics) = @_;

  # create quoted topic string
  my $topic_string = join(',', map(b($_)->quote, grep { $_ } @$topics));

  # Get topics and associated secrets
  my $db_request = <<"SELECT_TOPICS";
SELECT
  topic, secret FROM PubSub
WHERE
  topic in ($topic_string)
  AND mode = "subscribe"
  AND pending = 0
  AND (
    lease_seconds is NULL
    OR (started + lease_seconds) <= date("now")
  )
SELECT_TOPICS

  my $dbh = $c->dbh;
  my $array = $dbh->selectall_arrayref($db_request);

  # Todo: Is the hub the one I subscribed to?

  my (%topics, $secret);

  # Iterate through all topics
  foreach (@$array) {

    # No secret needed
    unless ($_->[1]) {

      # Topic is valid
      $topics{$_->[0]} = 1;
    }

    # Secret needed
    else {

      # No secret given
      if (!$secret) {

	# Init secret
	$secret = $_->[1] if $_->[1];
      }

      # Secret already given and mismatched for bulk
      elsif ($secret ne $_->[1]) {
	$c->app->log->info(
	  "Hub for topics $topic_string expects " .
	  'different secrets for bulk.');
	next;
      };

      # Secret matches for bulk
      $topics{$_->[0]} = 1;
    };
  };

  # Return filtered topics and secret
  return ([keys %topics], $secret);
});


# Add verification callback
app->callback(pubsub_verify => sub {
  my ($c, $params) = @_;

  my $dbh = $c->dbh;

  # Get subsrciption
  my $subscr = $dbh->selectrow_hashref(
    'SELECT FROM PubSub WHERE topic = ? AND mode = ? AND verify_token = ?',
    @{$params}{qw/topic mode verify_token/}
  );

  # No subscription of this topic found
  return unless $subscr;

  $dbh->begin_work;

  # Is subscription time over?
  if ($subscr->{lease_seconds} &&
	(time > ($subscr->{started} + $subscr->{lease_seconds}))) {

    # Delete Subscription (Maybe too hard?)
    unless ($dbh->do('DELETE FROM PubSub WHERE id = ?', $subscr->{id})) {
      $dbh->rollback and return;
    };
  };

  # If mode is subscribe and pending, update pending status

  # Maybe
  if ($subscr->{pending} && $subscr->{mode} eq 'subscribe') {
    unless ($dbh->do(
      'UPDATE PubSub SET pending = 0 WHERE id = ?', $subscr->{id}
    )) {
      $dbh->rollback and return;
    }
  }

  # If mode is unsubscribe, delete subscription
  # Maybe this is wrong?
  elsif ($subscr->{mode} eq 'unsubscribe') {

    # Delete subscription
    unless ($dbh->do('DELETE FROM PubSub WHERE id = ?', $subscr->{id})) {
      $dbh->rollback and return;
    };
  };

  # Everything is fine
  $dbh->commit;

  # Verify subscription
  return 1;
});


# How to handle the newly arrived content?
hook on_pubsub_content => sub {
  my ($c, $type, $dom) = @_;

  my (@feed, $elem);

  # Feed is Atom
  if ($type eq 'atom') {

    $elem = $dom->at('author > name');
    my $author = $elem ? $elem->all_text : undef;

    $dom->find('entry')->each(
      sub {
	my $entry = shift;

	my %info = (
	  topic => $entry->at('source > link[rel="self"]')->attrs('href')
	);

	foreach (qw/title id updated content/) {
	  $elem = $entry->at($_);
	  $_ = 'guid' if $_ eq 'id';
	  $info{$_} = $elem ? $elem->all_text : '';
	};

	$elem = $entry->at('author entry');
	$info{author} = $elem->all_text || $author;
	$info{updated} = $info{updated} ?
	  XML::Loy::Date::RFC3339->new( $info{updated} )->epoch
	      : time;

	push(@feed, \%info);
      }
    );
  }

  # Feed is RSS
  elsif ($type eq 'rss') {

    $dom->find('item')->each(
      sub {
	my $entry = shift;

	my %info = (
	  topic => $entry->at('source > link[rel="self"]')->attrs('href')
	);

	foreach (qw/title guid pubDate author description/) {
	  $elem = $entry->at($_);

	  # Rename pubDate
	  if ($_ eq 'pubDate') {
	    $_ = 'updated';
	  }

	  # Rename description
	  elsif ($_ eq 'description') {
	    $_ = 'content';
	  };

	  $info{$_} = $elem ? $elem->all_text : '';
	};

	# Set updated to epoch time
	$info{updated} =
	  $info{updated} ? Mojo::Date->new( $info{updated} )->epoch : time;

	push(@feed, \%info);
      });
  };

  my $dbh = $c->dbh;
  my $sth = $dbh->prepare(
    'INSERT INTO Content ' .
      '(topic, author, updated, title, content) ' .
	'VALUES ' .
	  '(?,?,?,?,?)'
  );

  # Start transaction
  $dbh->begin_work;

  # Import all entries to database
  foreach my $entry (@feed) {
    $sth->execute(@{$entry}{qw/topic author updated title content/});
  };

  # Commit insertions
  $dbh->commit;
};


# before_pubsub_(un)?subscribe
sub _store_subscription {
  my ($c, $param, $post) = @_;

  my %cond;

  my @first  = qw(hub lease_seconds secret verify_token mode pending started);
  my @second = qw(topic mode);

  $param->{pending} = 1;
  $param->{started} = time;

  my @values = (@{$param}{@first}, @{$param}{@second});

  my $sql =
    'UPDATE PubSub SET ' .
      join(', ', map { $_ . ' = ?' } @first) .
	' WHERE ' .
	  join(' AND ', map { $_ . ' = ?' } @second);

  my $dbh = $c->dbh;

  my $sth = $dbh->prepare($sql);

  return unless $sth;

  unless ($dbh->execute(@values)) {
    $sth = $dbh->prepare(
      'INSERT INTO PubSub (' .
	join(',', @first, @second) . ') VALUES (' . ('?' x 7) . ')'
      );
    unless ($dbh->execute(@values)) {
      app->log->warn('Unable to ' . $param->{mode} . ' to ' . $param->{topic});
    };
  };
  return;
};


# Unified event for subscription and unsubscribing
hook before_pubsub_subscribe => \&_store_subscription;


# Unified event for subscription and unsubscribing
hook before_pubsub_unsubscribe => \&_store_subscription;


##############
# Set routes #
##############
(any '/ps-callback')->pubsub;

# Show last content and subscription form
get '/' => sub {
  my $c = shift;
  # Todo: Get all feeds I subscribed to
  # Todo: Get latest entries
  $c->render(
    template => 'index'
  );
};

# Subscribe to new feed
post '/' => sub {
  my $c = shift;

  my $hub    = $c->param('hub');
  my $feed   = $c->param('feed');
  my $secret = $c->param('secret');

  # Missing information
  unless ($hub && $feed) {

    # Set information to flash
    $c->flash(
      hub    => $hub,
      feed   => $feed,
      secret => $secret
    );

    # Retry
    return $c->redirect_to('/');
  };

  # Create new parameter hash
  my %new_param = (
    topic => $feed,
    hub   => $hub
  );

  # Set secret
  $new_param{secret} = $secret if $secret;

  # Subscribe to new feed
  if ($c->pubsub_subscribe( %new_param )) {
    $c->flash(message => 'You subscribed to ' . $feed);
  }

  # Failed to subscribe to new feed
  else {
    $c->flash(message => 'Unable to subscribe to ' . $feed);
  };

  # Redirect
  return $c->redirect_to('/');
};


#######################
# Initialize Database #
#######################
sub _init_db {
  my $file = shift;

  my $dbh = DBI->connect("dbi:SQLite:dbname=$file", '', '');

  $dbh->begin_work;

  # Topic subscription
  unless ($dbh->do(
    'CREATE TABLE PubSub (
       id            INTEGER PRIMARY KEY,
       topic         TEXT NOT NULL,
       mode          TEXT NOT NULL,
       hub           TEXT,
       pending       INTEGER,
       lease_seconds INTEGER,
       secret        TEXT,
       verify_token  TEXT,
       started       INTEGER
     )'
  )) {
    $dbh->rollback and return;
  };

  # Topic subscription indices
  unless ($dbh->do(
    'CREATE INDEX IF NOT EXISTS pubsub_topic_i on PubSub (topic)'
  )) {
    $dbh->rollback and return;
  };

  # Content
  unless ($dbh->do(
    'CREATE TABLE PubSub_content (
       id       INTEGER PRIMARY KEY,
       author   TEXT,
       guid     TEXT,
       title    TEXT,
       updated  INTEGER,
       content  TEXT
     )')) {
    $dbh->rollback and return;
  };

  # Content indices
  foreach (qw/guid updated/) {
    unless ($dbh->do(
      "CREATE INDEX IF NOT EXISTS pubsub_content_${_}_i on PubSub_content (${_})"
    )) {
      $dbh->rollback and return;
    }
  };

  # Everything went fine
  $dbh->commit and return 1;
};

app->start;

__DATA__

@@ layouts/index.html.ep
<!doctype html>
<html>
  <head>
    <title><%= $title %></title>
  </head>
  <body>
    <h1><%= $title %></h1>
<%== content %>
  </body>
</html>

@@ index.html.ep
% layout 'index', title => 'Add Hub';

% my $subs = stash('subscriptions') || [];
% my $content = stash('entries') || [];

<h1>New Subscription</h1>
<form method="post" action="/">
% foreach (qw/feed hub secret/) {
  <label for="<%= $_ %>"><%= ucfirst($_) %></label>
  <input type="text" name="<%= $_ %>" id="<%= $_ %>" value="<%= flash($_) %>" />
  <br />
% }
  <input type="submit" value="OK" />
</form>

<h1>Subscriptions</h1>
% foreach my $sub ( @$subs ) {
<p>
  <%= $sub->{topic} %>
  <a hre="/unsubscribe/<%= $sub->{id} %>">unsubscribe</a>
</p>
% };

<h1>Content</h1>
% foreach my $entry ( @$content ) {
<h2><%= $entry->{title} %></h2>
<p><%= substr($entry->{content},200) %></p>
<p style="font-size: 60%">von <%= $entry->{author} %>,

% my $date = $entry->{updated} ?
%    XML::Loy::Date::RFC3339->new($entry->{updated})->to_string : '';

   <%= $date  %></p>
% };
